x500/cert/acme/src/main/java/org/wildfly/security/certificate/management/x500/cert/acme/AsyncAcmeClientSpi.java

```diff
@@ -0,0 +1,282 @@
+                });
+    }
+
+    private Uni<URL> getResourceUrl(AcmeAccount account, AcmeResource resource, boolean staging) {
+        return getResourceUrls(account, staging).map((Map<AcmeResource, URL> e) -> {
+            URL resourceUrl = e.get(resource);
+            if (resourceUrl == null) {
+                throw new RuntimeException(acme.resourceNotSupportedByAcmeServer(resource.getValue()));
+            }
+            return resourceUrl;
+        }).onFailure().transform(Throwable::getCause);
+    }
+
+    private static byte[] getReplayNonce(HttpResponse<String> connection) throws AcmeException {
+        String nonce = null;
+        if (connection.headers().firstValue(REPLAY_NONCE).isPresent()) {
+            nonce = connection.headers().firstValue(REPLAY_NONCE).get();
+        }
+        if (nonce == null) {
+            return null;
+        }
+        return CodePointIterator.ofString(nonce).base64Decode(BASE64_URL, false).drain();
+    }
+
+    public Uni<Map<AcmeResource, URL>> getResourceUrls(AcmeAccount account, boolean staging) {
+        if (account == null) {
+            return Uni.createFrom().failure(new IllegalArgumentException("account"));
+        }
+        final Map<AcmeResource, URL> resourceUrls = new HashMap<>();
+//        final Map<AcmeResource, URL> resourceUrls = account.getResourceUrls(staging);
+
+        Uni<HttpResponse<String>> httpResponseUni = null;
+        // TODO do we want caching?
+//        if (!resourceUrls.isEmpty()) {
+//            return Uni.createFrom().item(resourceUrls);
+//        }
+
+        if (staging && account.getServerUrl(true) == null) {
+            return Uni.createFrom().failure(acme.noAcmeServerStagingUrlGiven());
+        }
+        httpResponseUni = sendGetRequest(account.getServerUrl(staging), HttpURLConnection.HTTP_OK, JSON_CONTENT_TYPE);
+
+        return httpResponseUni.map((HttpResponse<String> httpResponse) -> {
+            JsonObject directoryJson;
+            try {
+                directoryJson = getJsonResponse(httpResponse);
+            } catch (AcmeException e) {
+                throw new RuntimeException(e);
+            }
+            for (AcmeResource resource : AcmeResource.values()) {
+                String resourceUrl = AcmeClientSpiUtils.getOptionalJsonString(directoryJson, resource.getValue());
+                URL url;
+                try {
+                    url = resourceUrl != null ? new URL(resourceUrl) : null;
+                } catch (MalformedURLException e) {
+                    throw new RuntimeException(acme.unableToRetrieveAcmeServerDirectoryUrls(e));
+                }
+                resourceUrls.put(resource, url);
+            }
+            return resourceUrls;
+        }).onFailure().transform(Throwable::getCause);
+    }
+
+    private Uni<HttpResponse<String>> sendGetRequest(String resourceUrl, int expectedResponseCode, String expectedContentType) {
+        HttpClient httpClient = HttpClient.newBuilder().build();
```

Comment:
Are we creating an instance of HTTP client per call?

Reply:
Good point, updated to not create new instance each time